c7n/resources/asg.py

```diff
@@ -1761,3 +1761,80 @@ def process_config(self, client, config):
             if e.response['Error']['Code'] == 'ValidationError':
                 return
             raise
+
+
+@resources.register('scaling-policy')
+class ScalingPolicy(query.QueryResourceManager):
+
+    class resource_type(query.TypeInfo):
+        service = 'autoscaling'
+        arn_type = "scalingPolicy"
+        id = name = 'PolicyName'
+        date = 'CreatedTime'
+        enum_spec = (
+            'describe_policies', 'ScalingPolicies', None
+        )
+        filter_name = 'PolicyNames'
+        filter_type = 'list'
+        cfn_type = 'AWS::AutoScaling::ScalingPolicy'
+
+
+@ASG.filter_registry.register('scaling-policy')
+class ScalingPolicyFilter(ValueFilter):
+
+    """Filter asg by scaling-policies attributes.
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: scaling-policies-with-target-tracking
+            resource: asg
+            filters:
+              - type: scaling-policy
+                key: PolicyType
+                value: "TargetTrackingScaling"
+
+    """
+
+    schema = type_schema(
+        'scaling-policy', rinherit=ValueFilter.schema
+    )
+    schema_alias = False
+    permissions = ("autoscaling:DescribePolicies",)
+
+    def initialize(self, asgs):
+        self.policies = self.get_scaling_policies(asgs)
+        return self
+
+    def get_scaling_policies(self, asgs):
+        policy_mgr = self.manager.get_resource_manager('scaling-policy')
+
+        policies = policy_mgr.resources()
+
+        if not policies:
+            return {}
+        policy_dict = {}
+        for policy in policies:
+            if policy['AutoScalingGroupName'] in policy_dict.keys():
+                policy_dict[policy['AutoScalingGroupName']].append(policy)
+            else:
+                policy_dict[policy['AutoScalingGroupName']] = [policy]
+
+        return policy_dict
+
+    def get(self, asg):
+        return self.policies.get(asg['AutoScalingGroupName'])
+
+    def process(self, asgs, event=None):
+        self.policy_info = self.initialize(asgs)
```

Comment:
initialize is returning self so this is creating a cycle memory reference, simply using the instance attribute dict of self.policies directly would be better.